url56b1dc76d3_wufunxiang_Libs_tgz-pJ8-org_msgpack_util_TemplatePrecompilerJ8/src/org/msgpack/template/builder/beans/XMLDecoder.java

```diff
@@ -0,0 +1,705 @@
+            }
+
+            if (toClose.isExpression) {
+                // push back expression
+                readObjs.push(toClose);
+            }
+        }
+
+        private Elem latestUnclosedElem() {
+            for (int i = readObjs.size() - 1; i >= 0; i--) {
+                Elem elem = readObjs.get(i);
+                if (!elem.isClosed) {
+                    return elem;
+                }
+            }
+            return null;
+        }
+
+        private Object execute(Elem elem) {
+            if (elem.isExecuted) {
+                return elem.result;
+            }
+
+            // execute to obtain result
+            try {
+                if (elem.idref != null) {
+                    elem.result = idObjMap.get(elem.idref);
+                } else if (elem.isBasicType) {
+                    elem.result = executeBasic(elem);
+                } else {
+                    elem.result = executeCommon(elem);
+                }
+            } catch (Exception e) {
+                listener.exceptionThrown(e);
+            }
+
+            // track id
+            if (elem.id != null) {
+                idObjMap.put(elem.id, elem.result);
+            }
+
+            elem.isExecuted = true;
+            return elem.result;
+        }
+
+        @SuppressWarnings("nls")
+        private Object executeCommon(Elem elem) throws Exception {
+            // pop args
+            ArrayList<Object> args = new ArrayList<Object>(5);
+            while (readObjs.peek() != elem) {
+                Elem argElem = readObjs.pop();
+                args.add(0, argElem.result);
+            }
+            // decide method name
+            String method = elem.methodName;
+            if (elem.fromProperty) {
+                method = (args.size() == 0 ? "get" : "set")
+                        + capitalize(method);
+            }
+            if (elem.fromIndex) {
+                Integer index = Integer.valueOf(method);
+                args.add(0, index);
+                method = args.size() == 1 ? "get" : "set";
+            }
+            if (elem.fromField) {
+                Field f = ((Class<?>) elem.target).getField(method);
+                return (new Expression(f, "get", new Object[] { null }))
+                        .getValue();
+            }
+            if (elem.fromOwner) {
+                return owner;
+            }
+
+            if (elem.target == owner) {
+                if ("getOwner".equals(method)) {
+                    return owner;
+                }
+                Class<?>[] c = new Class[args.size()];
+                for (int i = 0; i < args.size(); i++) {
+                    Object arg = args.get(i);
+                    c[i] = (arg == null ? null: arg.getClass());
+                }
+
+                // Try actual match method
+                try {
+                    Method m = owner.getClass().getMethod(method, c);
+                    return m.invoke(owner, args.toArray());
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

Catch `InvocationTargetException` explicitly when you call `Method.invoke()`. Use `Throwable.getCause()` or `getTargetException()` in the catch handler to access more information about the underlying exception. See the following diffs on GitHub that make this code change: [diff1](https://github.com/EsfingeFramework/gamification/commit/20792047e6c69457cca03e877f2ada578188f32a#diff-6808c12fe9ddaea042db84f95dd5ba78adab99d0cb18bb4d85916a8723530768L21), [diff2](https://github.com/cucumber-attic/cuke4duke/commit/f4821d400cf2590126f1298a8dff9f7a33d45dc5#diff-80adaffc20c0e6e7d2e42312386ffe1ddc10213bf5b8e8919895fd85d666c8deL34)

Learn more: [InvocationTargetException in the Java Platform, Standard Edition 8 API Specification](https://docs.oracle.com/javase/8/docs/api/java/lang/reflect/InvocationTargetException.html)